phoenix-core/src/main/java/org/apache/phoenix/coprocessor/PhoenixTTLRegionObserver.java

```diff
@@ -0,0 +1,368 @@
+                        + "[%s], scan = [%s], PHOENIX_TTL = %d ***************, numRequests=%d",
+                s.getRegionInfo().getTable().getNameAsString(), scan.toJSON(Integer.MAX_VALUE),
+                ScanUtil.getPhoenixTTL(scan), metricSource.getMaskExpiredRequestCount()));
+        return new PhoenixTTLRegionScanner(c.getEnvironment(), scan, s);
+    }
+
+    /**
+     * A region scanner that checks the TTL expiration of rows
+     */
+    private static class PhoenixTTLRegionScanner implements RegionScanner {
+        private static final String MASK_PHOENIX_TTL_EXPIRED_REQUEST_ID_ATTR =
+                "MASK_PHOENIX_TTL_EXPIRED_REQUEST_ID";
+
+        private final RegionScanner scanner;
+        private final Scan scan;
+        private final byte[] emptyCF;
+        private final byte[] emptyCQ;
+        private final Region region;
+        private final long minTimestamp;
+        private final long maxTimestamp;
+        private final long now;
+        private final boolean deleteIfExpired;
+        private final boolean maskIfExpired;
+        private final long requestId;
+        private long numRowsExpired;
+        private long numRowsScanned;
+        private long numRowsDeleted;
+        private boolean reported = false;
+
+        public PhoenixTTLRegionScanner(RegionCoprocessorEnvironment env, Scan scan,
+                RegionScanner scanner) throws IOException {
+            this.scan = scan;
+            this.scanner = scanner;
+            byte[] requestIdBytes = scan.getAttribute(MASK_PHOENIX_TTL_EXPIRED_REQUEST_ID_ATTR);
+            this.requestId = Bytes.toLong(requestIdBytes);
+
+            deleteIfExpired = ScanUtil.isDeleteTTLExpiredRows(scan);
+            maskIfExpired = !deleteIfExpired && ScanUtil.isMaskTTLExpiredRows(scan);
+
+            region = env.getRegion();
+            emptyCF = scan.getAttribute(EMPTY_COLUMN_FAMILY_NAME);
+            emptyCQ = scan.getAttribute(EMPTY_COLUMN_QUALIFIER_NAME);
+            byte[] txnScn = scan.getAttribute(BaseScannerRegionObserver.TX_SCN);
+            if (txnScn != null) {
+                TimeRange timeRange = scan.getTimeRange();
+                scan.setTimeRange(timeRange.getMin(), Bytes.toLong(txnScn));
+            }
+            minTimestamp = scan.getTimeRange().getMin();
+            maxTimestamp = scan.getTimeRange().getMax();
+            now = maxTimestamp != HConstants.LATEST_TIMESTAMP ?
+                            maxTimestamp :
+                            EnvironmentEdgeManager.currentTimeMillis();
+        }
+
+        @Override public int getBatch() {
+            return scanner.getBatch();
+        }
+
+        @Override public long getMaxResultSize() {
+            return scanner.getMaxResultSize();
+        }
+
+        @Override public boolean next(List<Cell> result) throws IOException {
+            try {
+                boolean hasMore;
```

Comment:
Extract this into a method since `nextRaw()` seems to be doing the exact same thing